File: clients/swift/Sources/messaging-cli/NetworkClient.swift
import Foundation

class NetworkClient {
    private let session: URLSession
    private let decoder: JSONDecoder
    private let encoder: JSONEncoder

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 10
        self.session = URLSession(configuration: config)
        self.decoder = JSONDecoder()
        self.encoder = JSONEncoder()
    }

    func register(username: String, password: String, serverURL: String) async throws -> RegisterResponse {
        let body = ["username": username, "password": password]
        let (data, response) = try await post(path: "/register", body: body, serverURL: serverURL, sessionId: nil)
        let status = (response as! HTTPURLResponse).statusCode
        if status == 409 { throw NetworkError.usernameTaken }
        if status == 401 { throw NetworkError.unauthorized }
        if status >= 400 { throw NetworkError.serverError(status) }
        return try decoder.decode(RegisterResponse.self, from: data)
    }

    func login(username: String, password: String, serverURL: String) async throws -> LoginResponse {
        let body = ["username": username, "password": password]
        let (data, response) = try await post(path: "/login", body: body, serverURL: serverURL, sessionId: nil)
        let status = (response as! HTTPURLResponse).statusCode
        if status == 401 { throw NetworkError.unauthorized }
        if status >= 400 { throw NetworkError.serverError(status) }
        return try decoder.decode(LoginResponse.self, from: data)
    }

    func send(to: String, text: String, sessionId: String, serverURL: String) async throws -> SendResponse {
        let body = ["to": to, "text": text]
        let (data, response) = try await post(path: "/send", body: body, serverURL: serverURL, sessionId: sessionId)
        let status = (response as! HTTPURLResponse).statusCode
        if status == 401 { throw NetworkError.unauthorized }
        if status >= 400 { throw NetworkError.serverError(status) }
        return try decoder.decode(SendResponse.self, from: data)
    }

    func getMessages(since: Int64, sessionId: String, serverURL: String) async throws -> MessagesResponse {
        guard var components = URLComponents(string: serverURL + "/messages") else {
            throw NetworkError.connectionFailed
        }
        components.queryItems = [URLQueryItem(name: "since", value: String(since))]
        guard let url = components.url else { throw NetworkError.connectionFailed }

        var request = URLRequest(url: url)
        request.setValue(sessionId, forHTTPHeaderField: "X-Session-Id")

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await session.data(for: request)
        } catch let error as URLError where error.code == .notConnectedToInternet {
            throw NetworkError.connectionFailed
        } catch {
            throw NetworkError.connectionFailed
        }

        let status = (response as! HTTPURLResponse).statusCode
        if status == 401 { throw NetworkError.unauthorized }
        if status >= 400 { throw NetworkError.serverError(status) }
        return try decoder.decode(MessagesResponse.self, from: data)
    }

    func logout(sessionId: String, serverURL: String) async {
        _ = try? await post(path: "/logout", body: [String: String](), serverURL: serverURL, sessionId: sessionId)
    }

    private func post(path: String, body: [String: String], serverURL: String, sessionId: String?) async throws -> (Data, URLResponse) {
        guard let url = URL(string: serverURL + path) else {
            throw NetworkError.connectionFailed
        }
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let sid = sessionId {
            request.setValue(sid, forHTTPHeaderField: "X-Session-Id")
        }
        request.httpBody = try encoder.encode(body)

        do {
            return try await session.data(for: request)
        } catch let error as URLError where error.code == .notConnectedToInternet {
            throw NetworkError.connectionFailed
        } catch {
            throw NetworkError.connectionFailed
        }
    }
}

File: clients/kotlin/src/main/kotlin/com/messaging/NetworkClient.kt
package com.messaging

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.net.ConnectException
import java.net.SocketException

class NetworkClient {
    private val http = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }
    private val JSON_TYPE = "application/json; charset=utf-8".toMediaType()

    suspend fun register(username: String, password: String, serverURL: String): RegisterResponse =
        withContext(Dispatchers.IO) {
            try {
                val body = json.encodeToString(RegisterRequest(username, password))
                val request = Request.Builder()
                    .url("$serverURL/register")
                    .post(body.toRequestBody(JSON_TYPE))
                    .build()
                http.newCall(request).execute().use { response ->
                    when (response.code) {
                        201 -> json.decodeFromString(response.body!!.string())
                        409 -> throw NetworkError.UsernameTaken
                        else -> throw NetworkError.ServerError(response.code)
                    }
                }
            } catch (e: NetworkError) {
                throw e
            } catch (e: ConnectException) {
                throw NetworkError.ConnectionFailed
            } catch (e: SocketException) {
                throw NetworkError.ConnectionFailed
            }
        }

    suspend fun login(username: String, password: String, serverURL: String): LoginResponse =
        withContext(Dispatchers.IO) {
            try {
                val body = json.encodeToString(LoginRequest(username, password))
                val request = Request.Builder()
                    .url("$serverURL/login")
                    .post(body.toRequestBody(JSON_TYPE))
                    .build()
                http.newCall(request).execute().use { response ->
                    when (response.code) {
                        200 -> json.decodeFromString(response.body!!.string())
                        401 -> throw NetworkError.Unauthorized
                        else -> throw NetworkError.ServerError(response.code)
                    }
                }
            } catch (e: NetworkError) {
                throw e
            } catch (e: ConnectException) {
                throw NetworkError.ConnectionFailed
            } catch (e: SocketException) {
                throw NetworkError.ConnectionFailed
            }
        }

    suspend fun send(to: String, text: String, sessionId: String, serverURL: String): SendResponse =
        withContext(Dispatchers.IO) {
            try {
                val body = json.encodeToString(SendRequest(to, text))
                val request = Request.Builder()
                    .url("$serverURL/send")
                    .header("X-Session-Id", sessionId)
                    .post(body.toRequestBody(JSON_TYPE))
                    .build()
                http.newCall(request).execute().use { response ->
                    when (response.code) {
                        200 -> json.decodeFromString(response.body!!.string())
                        401 -> throw NetworkError.Unauthorized
                        else -> throw NetworkError.ServerError(response.code)
                    }
                }
            } catch (e: NetworkError) {
                throw e
            } catch (e: ConnectException) {
                throw NetworkError.ConnectionFailed
            } catch (e: SocketException) {
                throw NetworkError.ConnectionFailed
            }
        }

    suspend fun getMessages(since: Long, sessionId: String, serverURL: String): MessagesResponse =
        withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$serverURL/messages?since=$since")
                    .header("X-Session-Id", sessionId)
                    .get()
                    .build()
                http.newCall(request).execute().use { response ->
                    when (response.code) {
                        200 -> json.decodeFromString(response.body!!.string())
                        401 -> throw NetworkError.Unauthorized
                        else -> throw NetworkError.ServerError(response.code)
                    }
                }
            } catch (e: NetworkError) {
                throw e
            } catch (e: ConnectException) {
                throw NetworkError.ConnectionFailed
            } catch (e: SocketException) {
                throw NetworkError.ConnectionFailed
            }
        }

    suspend fun logout(sessionId: String, serverURL: String): Unit =
        withContext(Dispatchers.IO) {
            try {
                val request = Request.Builder()
                    .url("$serverURL/logout")
                    .header("X-Session-Id", sessionId)
                    .post("{}".toRequestBody(JSON_TYPE))
                    .build()
                http.newCall(request).execute().use { }
            } catch (e: Exception) {
                // best-effort
            }
        }
}
